File: queue/queue.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

/*
Things to add:
    - Generics (make a generic queue)
    - ArrayQueue implementation
*/

class Node {
public:
    Node(string new_word) { word = new_word; }
    string word;
    Node* next;
};

class LinkQueue {
    Node* head;
    Node* last;
public:
    LinkQueue() {
        last = NULL;
        head = NULL;
    }

    bool is_empty() {
        if (head == NULL) {
            return true;
        } else {
            return false;
        }
    }

    string dequeue() {
        string word;

        if (head != NULL) {
            word = head->word;
            head = head->next;

            cout << "Dequeue: " << word << endl;
        } else {
            cout << "Empty queue" << endl;
        }
        return word;
    }

    void enqueue(string word) {
        Node* temp = new Node(word);
        temp->next = NULL;

        if (last != NULL)
            last->next = temp;

        if (head == NULL)
            head = temp;

        last = temp;
    }
};

int main() {
    LinkQueue* queue = new LinkQueue();
    string word;

    while (cin >> word) {
        // words that are '-' dequeue the first item put in the queue
        if (word == "-") {
            queue->dequeue();
        } else {
            queue->enqueue(word);
        }
    }

    return 0;
}

File: go/queue/queue.go
package main

import "fmt"

type node struct {
	 prev *node
	 next *node
	 data int
}

type queue struct {
	head *node
	tail *node
}

func (q *queue) enqueue(item int) {
	n := node{
		data: item,
		next: q.head,
		prev: nil,
	}
	if n.next == nil {
		q.tail = &n
	} else {
		n.next.prev = &n
	}
	q.head = &n
}

func (q *queue) dequeue() (int, bool) {
	if q.tail != nil {
		rec := q.tail
		if rec.prev != nil {
			q.tail = rec.prev
			rec.prev.next = nil
		} else {
			q.tail = nil
		}
		return rec.data, true
	} else {
		return 0, false
	}
}

func main() {
	var q queue
	q.enqueue(1)
	q.enqueue(2)
	q.enqueue(3)
	q.enqueue(4)
	q.enqueue(5)
	q.enqueue(6)
	fmt.Println(q.dequeue())
	fmt.Println(q.dequeue())
	fmt.Println(q.dequeue())
	fmt.Println(q.dequeue())
	fmt.Println(q.dequeue())
	fmt.Println(q.dequeue())
	fmt.Println(q.dequeue())
}